### PR TITLE
[low-power] add receive timing capability to CSL receiver

### PR DESCRIPTION
**Pull request by edmont**
This PR uses the receive timing capability from the radio to schedule CSL receive operations with a higher precision, allowing for shorter receive windows.

`mCslTimer` is set up to be triggered `OPENTHREAD_CONFIG_CSL_RECEIVE_TIME_AHEAD` before the start of the receive window.

Depends on:
- [x]  https://github.com/openthread/openthread/pull/6311 [low-power] adaptative CSL window duration

### CODE REVIEW

**Comment by edmont** [comment]
In the attached capture CSL transmissions are being triggered 300 s after last synchronization. Build from this PR is configured with ±1 ppm crystals in both ends and ±100 us parent uncertainty. CSL receive window duration goes from 492 us at the sync time to 1692 us after 300 s.

[csl_tx_300s.pcap.zip](https://github.com/openthread/openthread/files/6213125/csl_tx_300s.pcap.zip)

**Comment by Irving-cl** [comment]
I'm thinking that, for SSED, not all 'Receive' operations are for CSL Sampling. For example, an SSED can also send a data poll and wait for packets. If we implement `otPlatRadioReceive` as:
```
#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
    if (sCslSampleDur > 0)
    {
        result = nrf_802154_receive_at(sCslSampleTime, sCslSampleStart, sCslSampleDur, aChannel);
    }
    else
#endif
```
In the data poll case, the SSED need to start receiving immediately after it sends the poll. It should't wailt until next sample window.

Thoughts? @edmont @brianljt @abtink

File: src/core/radio/radio.hpp [hunk 1]
```diff
@@ -408,11 +408,22 @@ class Radio : public InstanceLocator, private NonCopyable
...
 
+#if OPENTHREAD_CONFIG_MAC_SOFTWARE_RX_TIMING_ENABLE
+    /**
+     * This method updates the CSL start time and duration in radio.
+     *
+     * @param[in]  aCslSampleStart  The CSL sample start time.
+     * @param[in]  aCslSampleDur    The CSL sample duration.
+     *
+     */
+    void UpdateCslSampleStart(uint32_t aCslSampleStart, uint32_t aCslSampleDur);
```
Comment (brianljt) [inline src/core/radio/radio.hpp:424]: Looks more like updating a sample window(start and end) than just start to me
```suggestion
    void UpdateCslSampleWindow(uint32_t aCslSampleStart, uint32_t aCslSampleDur);
```

File: examples/platforms/nrf528xx/src/radio.c [hunk 2]
```diff
@@ -489,7 +491,16 @@ otError otPlatRadioReceive(otInstance *aInstance, uint8_t aChannel)
 
     nrf_802154_tx_power_set(GetTransmitPowerForChannel(aChannel));
 
-    result = nrf_802154_receive();
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
+    if (sCslSampleDur > 0)
```
Comment (brianljt) [inline examples/platforms/nrf528xx/src/radio.c:495]: I would suggest to use a separate API like `otPlatRadioReceiveAt` for this purpose rather than reuse `otPlatRadioReceive` with some global parameters. There are some wakes not related to CSL managed by OT stack. If we change this implementation like this, it would mess up the OT managed wake.

Comment (edmont) [reply examples/platforms/nrf528xx/src/radio.c:495]: Makes sense, I'll update.

File: src/core/mac/sub_mac.cpp [hunk 3]
```diff
@@ -1013,16 +1012,25 @@ void SubMac::HandleCslTimer(void)
         break;
 
     case kCslSleep:
-        mCslState = kCslSample;
+        mCslSampleTime += periodUs;
 
+#if OPENTHREAD_CONFIG_MAC_SOFTWARE_RX_TIMING_ENABLE
```
Comment (brianljt) [inline src/core/mac/sub_mac.cpp:1017]: Define something like `OT_RADIO_CAPS_RECEIVE_TIMING` like `OT_RADIO_CAPS_TRANSMIT_TIMING` as radio capability and replace this macro check with runtime check `SubMac::ShouldHandleReceiveTargetTime` like `SubMac::ShouldHandleTransmitTargetTime`?

Comment (edmont) [reply src/core/mac/sub_mac.cpp:1017]: This was my original idea for the first commit (https://github.com/openthread/openthread/pull/6343/commits/0d92ffb0cd52937527776e681b7fb92a46b18236), but I see two problems:
- Adds unneeded code size.
- Forces all platforms to implement a dummy `otPlatRadioReceiveAt`.

Comment (brianljt) [reply src/core/mac/sub_mac.cpp:1017]: I think it is for supporting both NCP mode and RCP mode(posix host + RCP firmware). In RCP mode, `SubMac.cpp` will be included in **both** posix host and RCP firmware. In host side, it will get the radio capability from `SubMac::GetCaps` which combines the real radio capability and the SW capability as the `SubMac` can add SW capability which is missed in radio. If `posix host` finds RCP supports `OT_RADIO_CAPS_RECEIVE_TIMING`, it should not start the receive timer and leave it to RCP. On RCP side, this could be done in `SubMac` layer or radio layer, depends on whether radio has the capability or not in `Get<Radio>().GetCaps()`. If radio supports it, it will leave it to radio and should not start receive timer, if not, it will handle it in `SubMac` layer.

In NCP mode, there will be only one `SubMac` layer and it decides if `SubMac` SW stack should handle the receive timer or not base on Radio's capability.

The slight increase of code size is unavoidable as we share the same `SubMac` code for both posix host build and NCP, RCP firmware build. This is also inline with other radio capabilities like `OT_RADIO_CAPS_ACK_TIMEOUT`, `OT_RADIO_CAPS_TRANSMIT_RETRIES` and `OT_RADIO_CAPS_CSMA_BACKOFF`.

For dummy `otPlatRadioReceiveAt`, I think you can define a weak function in general in `src/core/radio/radio_platform.cpp` to avoid implement it for all platforms like what we have for `otPlatRadioSetMacFrameCounter` or `otPlatRadioGetCslAccuracy`. How do you think?

Comment (edmont) [reply src/core/mac/sub_mac.cpp:1017]: Thanks for the clarification. It makes sense, I'll add the changes.

File: examples/platforms/nrf528xx/src/radio.c [hunk 4]
```diff
@@ -488,13 +489,27 @@ otError otPlatRadioReceive(otInstance *aInstance, uint8_t aChannel)
...
 
     nrf_802154_tx_power_set(GetTransmitPowerForChannel(aChannel));
-
     result = nrf_802154_receive();
     clearPendingEvents();
 
     return result ? OT_ERROR_NONE : OT_ERROR_INVALID_STATE;
 }
 
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE && OPENTHREAD_CONFIG_MAC_SOFTWARE_RX_TIMING_ENABLE
```
Comment (brianljt) [inline examples/platforms/nrf528xx/src/radio.c:498]:
```suggestion
#if OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2
```
`otPlatRadioGetCaps` indicates `OT_RADIO_CAPS_RECEIVE_TIMING` for OT1.2, then this method should be provided for OT1.2.

File: examples/platforms/nrf528xx/src/radio.c [hunk 5]
```diff
@@ -1065,7 +1081,17 @@ void nrf_802154_receive_failed(nrf_802154_rx_error_t error)
     sAckedWithSecEnhAck = false;
 #endif
 
-    setPendingEvent(kPendingEventReceiveFailed);
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE && OPENTHREAD_CONFIG_MAC_SOFTWARE_RX_TIMING_ENABLE
```
Comment (brianljt) [inline examples/platforms/nrf528xx/src/radio.c:1084]:
```suggestion
#if OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2
```

**Review by brianljt — APPROVED** [review]
LGTM 👍, Thanks!

**Review by Irving-cl — APPROVED** [review]
LGTM

**Comment by edmont** [comment]
Added last review changes and squashed.

File: include/openthread/platform/radio.h [hunk 6]
```diff
@@ -680,6 +681,18 @@ otError otPlatRadioSleep(otInstance *aInstance);
  */
 otError otPlatRadioReceive(otInstance *aInstance, uint8_t aChannel);
 
+/**
+ * Schedule a radio reception window at a specific time and duration.
+ *
+ * @param[in]  aChannel  The radio channel on which the frame is to be received.
+ * @param[in]  aStart     The receive window start time, in microseconds.
```
Comment (jwhui) [inline include/openthread/platform/radio.h:688]: nit: alignment
```suggestion
 * @param[in]  aChannel   The radio channel on which the frame is to be received.
 * @param[in]  aStart     The receive window start time, in microseconds.
```

File: examples/platforms/nrf528xx/nrf52811/openthread-core-nrf52811-config.h [hunk 7]
```diff
@@ -160,6 +160,16 @@
...
 #endif
 
+/**
+ * @def OPENTHREAD_CONFIG_MAC_SOFTWARE_RX_TIMING_ENABLE
+ *
+ * Define to 1 to enable software reception target time logic.
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_MAC_SOFTWARE_RX_TIMING_ENABLE
+#define OPENTHREAD_CONFIG_MAC_SOFTWARE_RX_TIMING_ENABLE (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
```
Comment (jwhui) [inline examples/platforms/nrf528xx/nrf52811/openthread-core-nrf52811-config.h:170]: Is it safe to use `OPENTHREAD_CONFIG_THREAD_VERSION` in the project config? The defaults are included after the project config.

https://github.com/openthread/openthread/blob/8467a3d5223c274b65a71080f6aad94dc182a9bb/src/core/openthread-core-config.h#L45-L51

File: src/core/radio/radio.hpp [hunk 8]
```diff
@@ -408,11 +408,24 @@ class Radio : public InstanceLocator, private NonCopyable
     /**
      * This method updates the CSL sample time in radio.
      *
-     * @param[in]  aCslSampleTime  The CSL sample time.
+     * @param[in]  aCslSampleTime   The CSL sample time.
```
Comment (jwhui) [inline src/core/radio/radio.hpp:411]: nit: unnecessary change?
```suggestion
     * @param[in]  aCslSampleTime  The CSL sample time.
```

File: src/core/radio/radio.hpp [hunk 9]
```diff
@@ -408,11 +408,24 @@ class Radio : public InstanceLocator, private NonCopyable
...
      *
      */
     void UpdateCslSampleTime(uint32_t aCslSampleTime);
 
+    /**
+     * This method schedules a radio reception window at a specific time and duration.
+     *
+     * @param[in]  aChannel   The radio channel on which the frame is to be received.
```
Comment (jwhui) [inline src/core/radio/radio.hpp:419]: nit: we don't know that there will be a frame to receive
```suggestion
     * @param[in]  aChannel   The radio channel on which to receive.
```

File: examples/platforms/nrf528xx/nrf52811/openthread-core-nrf52811-config.h [hunk 10]
```diff
@@ -226,6 +236,28 @@
...
+ *
+ */
+#ifndef OPENTHREAD_CONFIG_CSL_RECEIVE_TIME_AHEAD
+#define OPENTHREAD_CONFIG_CSL_RECEIVE_TIME_AHEAD 2000
+#endif
+
+/**
+ * @def OPENTHREAD_CONFIG_CSL_MIN_RECEIVE_ON
+ *
+ * The total duration, in units of microseconds, required for the CSL receiver to fully receive and acknowledge a frame
```
Comment (jwhui) [inline examples/platforms/nrf528xx/nrf52811/openthread-core-nrf52811-config.h:252]: I think this is something more like: "The minimum CSL receive window (in microseconds) required to receive an IEEE 802.15.4 frame."

Total duration indicates the total time it takes to receive and transmit the ack, which should be much longer than the 12 * 16 = 196us value used below.

File: examples/platforms/nrf528xx/nrf52811/openthread-core-nrf52811-config.h (hunk 7 above)
Comment (edmont) [reply examples/platforms/nrf528xx/nrf52811/openthread-core-nrf52811-config.h:170]: It seems not. This was just inheriting convention from previous TX capabilities.

I would say the right way should be to just confirm or deny the capabilities in the platforms configs (`0` or `1`) and make sure they are used correctly within OT code depending on the configured Thread version.

But could we address this in a different PR?

Comment (jwhui) [reply examples/platforms/nrf528xx/nrf52811/openthread-core-nrf52811-config.h:170]: I am OK to resolve in a separate PR.

**Review by jwhui — APPROVED** [review]
